### ISSUE TITLE
Support to analyze DynamoDB streaming configuration

### ISSUE THREAD

**Issue by iainelder** [jckuester/awsls#35]
I'm using awsls to analyze the streaming configuration for a set of DynamoDB tables.

There are two types of stream to analyze:

* DynamoDB streams (awls already supports this)
* Kinesis Data Streams for DynamoDB (awsls has no support for this)

I want this because the DynamoDB API doesn't provide a simple way to get information about tables in bulk.

I was trying to avoid writing a bash script around the AWS CLI (which is fragile) or writing a Python script around the boto3 library (which can be a lot of extra work because of how it handles things like pagination).

### Kinesis Data Streams for DynamoDB

The [Terraform AWS provider version 3.39.0](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md#3390-may-06-2021) introduces the resource  aws_dynamodb_kinesis_streaming_destination.

awls version 0.10.0 appears to be using provider version 3.31.0, which doesn't have the resource. So the following command fails.

```
$ awsls --debug --json --profiles ... --regions ... aws_dynamodb_kinesis_streaming_destination

Error: no resource type found: aws_dynamodb_kinesis_streaming_destination
```

**Would upgrading the provider be enough to support this?**

I'm asking because I'm not sure if awsls in theory supports all Terraform resources or only those that also have a data version. 

### DynamoDB streams

As of version 0.10.0, awls already solves the problem beautifully for DynamoDB streams. (Thanks to @9to6 for the JSON support!)

```text
$ time awsls --json --debug --profiles dan-2nd --regions eu-west-1 aws_dynamodb_table \
--attributes stream_enabled,stream_view_type,stream_arn,stream_label \
> ~/tmp/table_streaming_configuration
...
real	0m33,518s
user	0m4,905s
sys	0m1,273s


$ cat ~/tmp/table_streaming_configuration | jq length
104


$ cat ~/tmp/table_streaming_configuration |
  jq 'map(select(.stream_enabled == "true" and (.id | startswith("prod-")))) | length'
14
```

**Comment by iainelder** [comment]
My current workaround is to iterate over the list of tables with xargs and describe the Kinesis streaming destinations one by one.

It works, but awsls is what I use to try to avoid this kind of scripting :-)

```
$ cat ~/tmp/table_streaming_configuration |
    jq -r 'map(select((.id | startswith("prod-"))) | .id)[]' |
    xargs -n1 aws dynamodb describe-kinesis-streaming-destination --profile ... --table-name |
    jq -s
[
  {
    "TableName": "prod-foobar",
    "KinesisDataStreamDestinations": []
  },
```

**Comment by jckuester** [comment]
Hi @iainelder, thanks for your resource request. I should upgrade `awsls` more frequently to support new resources :)

Here is PR for now: https://github.com/jckuester/awsls/pull/36/. With `go build` you can create a temporary binary until the release.

> I'm asking because I'm not sure if awsls in theory supports all Terraform resources or only those that also have a data version.

What do you mean by data version? In theory, awsls supports all Terraform resources, yes. In practice it is a little bit more difficult: awsls is mostly code-generated (the generator is in the folder `gen/`) and there a few edge-cases to generate code that covers all resources (I am still working on that). However, to rerun the code generator whenever the Terraform AWS Provider released new resources (like in this case), is not that difficult.

**Comment by iainelder** [comment]
Hi @jckuester. Thanks for the response. I'll test that PR when I have a spare moment.

By data version I mean that some of the AWS resource types have read-only versions implemented as data source.

For example, 

* [DynamoDB table data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/dynamodb_table)
* [DynamoDB table resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table)

But you've made it clear in your response that all resources are supported. That's good because not all resources have data sources!

**Comment by iainelder** [comment]
Finally found some time to test the PR. Sorry for the delay!

I compiled the branch in #36 and repeated the command to list the aws_dynamodb_kinesis_streaming_destination resources.

It still fails, but there is an extra output in the debug message saying that it is "not (yet) supported".

```text
$ ./awsls --debug --json --profiles ... --regions ... aws_dynamodb_kinesis_streaming_destination

   • resource type not (yet) supported: aws_dynamodb_kinesis_streaming_destination
Error: no resource type found: aws_dynamodb_kinesis_streaming_destination
```

**Comment by iainelder** [comment]
It turns out that aws_dynamodb_kinesis_streaming_destination is not present in the list of SupportedTypes.

https://github.com/jckuester/awsls/blob/1855e7b59e6e9b00298a14903a7f38c43fb6b208/resource/supported.go

It's not clear to me how this list is generated.

Does something else need to be changed to support this resource?

**Comment by jckuester** [comment]
Hi @iainelder,

I just realised that `aws_dynamodb_kinesis_streaming_destination` has a required attribute (the [`TableName`](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DescribeKinesisStreamingDestination.html)) and `awsls` can currently not list resources that have required attributes. 

I am planning to support resources with required parameters, but I haven't yet thought through how to design this properly. 

The default case would probably be to list the `aws_dynamodb_kinesis_streaming_destination` resources for all tables? However, I am not sure if listing all resources that have required attributes when running `awsls "*` is the desired behavior; because awsls would return for resources such as `aws_s3_bucket_object` massive amount of items/keys in all your s3 buckets....

**Comment by iainelder** [comment]
Interesting! Now it starts to make sense why not all of the Terraform resources are supported.

Does awsls make direct API calls to AWS? I was under the impression that the awsls was using the only Terraform APIs to read resources. But I'm still unfamiliar with the Terraform API, so I don't know how that would work.

> The default case would probably be to list the aws_dynamodb_kinesis_streaming_destination resources for all tables?

In this case there is at most one streaming destination for each table, so it seems practical to return all the resources by default. I imagine this would work as in my Bash-based solution. First all the tables would be listed. Then the streaming destination resource would be retrieved for each one. If the table has no streaming destination then it produces no output (but it might produce a debug message).

> However, I am not sure if listing all resources that have required attributes when running awsls "*" is the desired behavior

I can see how that could be a problem for resources like aws_s3_bucket_object and aws_dynamodb_table_item. I think in this case the correct default behavior is for the glob to ignore them. If they need to be included at all, then they should be named explicitly.

You could follow the precedent set by CloudTrail. It has the same problem when logging API calls. The calls to create S3 buckets and their configurations are infrequent and are logged by default as "management events". But the calls to access S3 objects for a "busy" bucket with thousands of objects being created per hour could overwhelm the logs. So S3 object logging is disabled by default. To enable it you must explicitly configure "data events".

In CloudTrail data events include S3 object-level API activity, DynamoDB item-level activity, and other data types which I am unfamiliar with in a Terraform context: https://docs.aws.amazon.com/awscloudtrail/latest/userguide/logging-data-events-with-cloudtrail.html

To be honest though, the best thing for now might be to not implement listings for S3 objects.

**Comment by jckuester** [comment]
> Does awsls make direct API calls to AWS? I was under the impression that the awsls was using the only Terraform APIs to read resources. 

Yes, `awsls` does call the AWS API directly. The code for it is generated; I adapted the existing aws-sdk-go code generator (see [here](https://github.com/aws/aws-sdk-go/tree/main/private)) for the `awsls` use case. Terraform AWS Provider API is only called for getting resources' attributes (tags, ARNs, etc.) in a consistent way. For example, for tags there are many different ways to get them via the AWS API, but when getting tags via the Terraform AWS Provider, this gets abstracted away for me.

Do you have any opinion/feedback about how required parameters could be provided from the user interface perspective?

I was thinking of `awsls aws_dynamodb_kinesis_streaming_destination --required 'aws_dynamodb_table.name=*'` or so...

**Comment by iainelder** [comment]
Would this return all the tables?

```
awsls aws_dynamodb_kinesis_streaming_destination --required 'aws_dynamodb_table.name=*'
```

To list all the streaming destinations this should just work:

```
awsls aws_dynamodb_kinesis_streaming_destination
```

At the API level there is a one-to-one correspondence between the TableName input and the the Kinesis streaming destination.

Even if that's how the API works I don't expect to see that complexity exposed in the user interface of the tool.

If I understand your syntax, then this command would return the streaming destination of Table1.

```
awsls aws_dynamodb_kinesis_streaming_destination --required 'aws_dynamodb_table.name=Table1'
```

Filtering resources is not supported for any of the other resources so I wouldn't expect that power here.

For me it's fine just to output all of the resources and I can filter them by myself with other tools.
